spec.md

```diff
@@ -0,0 +1,1597 @@
+* **Generalizes** to $n$ inputs and $m$ outputs
+
+**Case SINGLE_RET_FUN**
+```
+fun f (x: T, y: S) -> A, B:
+  match <PATTERN>
+  (<OPERATORS>)
+  return <AGG>, <AGG>;
+```
+adds the following to our type system:
+* _Function symbol_: when $`x : T`$ and $y: S$ then $f(x:T, y:S) : A \times B$
+* _Function terms_: $`(z,w) : f(x:T, y:S)`$ are discussed in section "Function semantics"
+* **Generalizes** to $n$ inputs and $m$ outputs
+
+_Comment: notice difference in capitalization between the two cases!_
+
+## Undefine semantics
+
+### Type axioms
+
+**Case ENT**
+* `entity A` removes $`A : \mathbf{Ent}`$
+* `sub B from (entity) A` removes $`A < B`$
+
+**Case REL**
+* `relation A` removes $`A : \mathbf{Rel}`$
+* `sub B from (relation) A` removes $`A < B`$
+* `relates I from (relation) A` removes $`A : \mathbf{Rel}(I)`$
+* `as J from (relation) A relates I` removes $`I <_! J`$ 
+* `relates I[] from (relation) A` removes $`A : \mathbf{Rel}([I])$
+* `as J[] from (relation) A relates I[]` removes $`I <_! J`$
+
+**Case ATT**
+* `attribute A` removes $`A : \mathbf{Att}`$ and $`A : \mathbf{Att}(O_A)`$
+* `value V from (attribute) A value V` removes $`A < V`$
+* `sub B from (attribute) A` removes $`A <_! B`$ and $`O_A <_! O_B`$
+
+**Case PLAYS**
+* `plays B:I from (type) A` removes $`A <_! I`$ 
+
+**Case OWNS**
+* `owns B from (type) A` removes $`A <_! O_B`$ 
+* `owns B[] from (type) A` removes $`A <_! O_B`$
+
+### Constraints
+
+_In each case, `undefine` removes the postulated condition (restoring the default)._ (minor exception: subkey)
+
+**Case CARD**
+* `@card(n..m) from A relates I`
+* `@card(n..m) from A plays B:I`
+* `@card(n...m) from A owns B`
+
+**Case CARD_LIST**
+* `@card(n..m) from A relates I[]`
+* `@card(n...m) from A owns B[]`
+
+**Case PLAYS_AS**
+* `as C from A plays B`
+
+**Case OWNS_AS**
+* `as C from A owns B`
+
+**Case UNIQUE**
+* `@unique from A owns B`
+
+**Case KEY**
+* `@key from A owns B`
+
+**Case SUBKEY**
+* `@subkey(<LABEL>) from A owns B` removes $`B`$ as part of the `<LABEL>` key of $`A`$
+
+**Case ABSTRACT**
+* `@abstract from (type) B` 
+* `@abstract from A plays B:I`
+* `@abstract from A owns B` 
+* `@abstract from A owns B[]` 
+* `@abstract from B relates I`
+* `@abstract from B relates I[]`
+
+**Case VALUES**
+* `@values(v1, v2) from A owns B` 
+* `@range(v1..v2) from A owns B`
+* `@values(v1, v2) from A values B` 
+* `@range(v1..v2) from A values B`
+
+**Case DISTINCT**
+* `@distinct from A owns B[]`
+* `@distinct from B relates I[]`
+
+### Triggers
+
+_In each case, `undefine` removes the triggered action._
+
+**Case DEP_DEL (CASCADE/INDEPEDENT)**
+* `@cascade from (relation) B relates I`
+* `@cascade from (relation) B`
+* `@independent from (attribute) B`
+
+### Value types
+
+**Case PRIMITIVES**
+cannot undefine primitives
+
+**Case STRUCT**
+
+* `struct S;`
+  removes $S : \mathbf{Type}$ and all associated defs.
+  * **TT** error if
```

Comment:
This is pretty good, but need more of this on previous undefine cases, illustrating errors thrown. Some sweeping statements could be made e.g. undefining a type if there are instances of that type (data-storing or interface).